transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/persistence/SQLiteEventStore.java

```diff
@@ -0,0 +1,355 @@
+
+              // TODO: insert all with one sql query.
+              for (Map.Entry<String, String> entry : event.getMetadata().entrySet()) {
+                ContentValues metadata = new ContentValues();
+                metadata.put("event_id", newEventId);
+                metadata.put("name", entry.getKey());
+                metadata.put("value", entry.getValue());
+                db.insert("event_metadata", null, metadata);
+              }
+              return newEventId;
+            });
+
+    return PersistedEvent.create(newRowId, backendName, event);
+  }
+
+  @Override
+  public void recordFailure(Iterable<PersistedEvent> events) {
+    StringBuilder query =
+        new StringBuilder("UPDATE events SET num_attempts = num_attempts + 1 WHERE _id in (");
+    Iterator<PersistedEvent> iterator = events.iterator();
+    while (iterator.hasNext()) {
+      query.append(iterator.next().getId());
+      if (iterator.hasNext()) {
+        query.append(',');
+      }
+    }
+    query.append(')');
+    inTransaction(
+        db -> {
+          db.compileStatement(query.toString()).execute();
+          db.compileStatement("DELETE FROM events WHERE num_attempts >= " + MAX_RETRIES).execute();
+          return null;
+        });
+  }
+
+  @Override
+  public void recordSuccess(Iterable<PersistedEvent> events) {
+    StringBuilder query = new StringBuilder("DELETE FROM events WHERE _id in (");
+    Iterator<PersistedEvent> iterator = events.iterator();
+    while (iterator.hasNext()) {
+      query.append(iterator.next().getId());
+      if (iterator.hasNext()) {
+        query.append(',');
+      }
+    }
+    query.append(')');
+    getDb().compileStatement(query.toString()).execute();
+  }
+
+  @Override
+  @Nullable
+  public Long getNextCallTime(String backendName) {
+    try (Cursor cursor =
+        getDb()
+            .rawQuery(
+                "SELECT next_request_ms FROM backends WHERE name = ?",
+                new String[] {backendName})) {
+      if (cursor.moveToNext()) {
+        return cursor.getLong(0);
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public boolean hasPendingEventsFor(String backendName) {
+    try (Cursor cursor =
+        getDb()
+            .rawQuery(
+                "SELECT 1 FROM events WHERE backend_id = ? LIMIT 1", new String[] {backendName})) {
+      return cursor.moveToNext();
+    }
+  }
+
+  @Override
+  public void recordNextCallTime(String backendName, long timestampMs) {
+    inTransaction(
+        db -> {
+          ContentValues values = new ContentValues();
+          values.put("next_request_ms", timestampMs);
+          int rowsUpdated = db.update("backends", values, "name = ?", new String[] {backendName});
+
+          if (rowsUpdated < 1) {
+            values.put("name", backendName);
+            db.insert("backends", null, values);
+          }
+          return null;
+        });
+  }
+
+  @Override
+  public Iterable<PersistedEvent> loadAll(String backendName) {
```

Comment:
This function is huge. Consider breaking it up into methods whose names document what you're doing.